src/components/js-tracer/Canvas.js

```diff
@@ -0,0 +1,568 @@
+  |},
+|};
+
+const TEXT_OFFSET_TOP: CssPixels = 11;
+const TEXT_OFFSET_START: CssPixels = 3;
+const ROW_LABEL_OFFSET_LEFT: CssPixels = 5;
+const FONT_SIZE: CssPixels = 10;
+
+class JsTracerCanvas extends React.PureComponent<Props, State> {
+  state: State = {
+    hasFirstDraw: false,
+  };
+
+  drawCanvas = (
+    ctx: CanvasRenderingContext2D,
+    hoveredItem: IndexIntoJsTracerEvents | null
+  ) => {
+    const {
+      rowHeight,
+      jsTracerTimingRows,
+      viewport: {
+        viewportTop,
+        viewportBottom,
+        containerWidth,
+        containerHeight,
+      },
+    } = this.props;
+
+    const { devicePixelRatio } = window;
+
+    // Set the font size before creating a text measurer.
+    ctx.font = `${FONT_SIZE * devicePixelRatio}px sans-serif`;
+
+    const renderPass: RenderPass = {
+      ctx,
+      textMeasurement: new TextMeasurement(ctx),
+      fastFillStyle: new FastFillStyle(ctx),
+      startRow: Math.floor(viewportTop / rowHeight),
+      endRow: Math.min(
+        Math.ceil(viewportBottom / rowHeight),
+        jsTracerTimingRows.length
+      ),
+      devicePixels: {
+        // Convert many of the common values provided by the Props into DevicePixels.
+        containerWidth: containerWidth * devicePixelRatio,
+        innerContainerWidth:
+          (containerWidth - TIMELINE_MARGIN_LEFT - TIMELINE_MARGIN_RIGHT) *
+          devicePixelRatio,
+        containerHeight: containerHeight * devicePixelRatio,
+        textOffsetStart: TEXT_OFFSET_START * devicePixelRatio,
+        textOffsetTop: TEXT_OFFSET_TOP * devicePixelRatio,
+        rowHeight: rowHeight * devicePixelRatio,
+        viewportTop: viewportTop * devicePixelRatio,
+        timelineMarginLeft: TIMELINE_MARGIN_LEFT * devicePixelRatio,
+        timelineMarginRight: TIMELINE_MARGIN_RIGHT * devicePixelRatio,
+        oneCssPixel: devicePixelRatio,
+        rowLabelOffsetLeft: ROW_LABEL_OFFSET_LEFT * devicePixelRatio,
+      },
+    };
+
+    {
+      // Clear out any previous events.
+      const { fastFillStyle, devicePixels } = renderPass;
+      fastFillStyle.set('#ffffff');
+      ctx.fillRect(
+        0,
+        0,
+        devicePixels.containerWidth,
+        devicePixels.containerHeight
+      );
+    }
+
+    this.drawEvents(renderPass, hoveredItem);
+    this.drawSeparatorsAndLabels(renderPass);
+
+    if (!this.state.hasFirstDraw) {
+      this.setState({ hasFirstDraw: true });
+    }
+  };
+
+  // Note: we used a long argument list instead of an object parameter on
+  // purpose, to reduce GC pressure while drawing.
+  drawOneEvent(
+    renderPass: RenderPass,
+    x: DevicePixels,
+    y: DevicePixels,
+    w: DevicePixels,
+    h: DevicePixels,
+    uncutWidth: DevicePixels,
+    text: string,
+    backgroundColor: string = BLUE_40,
+    foregroundColor: string = 'white'
+  ) {
+    const { ctx, textMeasurement, fastFillStyle, devicePixels } = renderPass;
+
+    fastFillStyle.set(backgroundColor);
+
+    if (uncutWidth >= 1) {
```

Comment:
optional because it may add too much complexity, but you may have a good idea: ideally I think we should enter this branch if the event fills 1 pixel entirely. This means that if `uncutWidth` is 1, but goes from 4.3 to 5.3, then there's no full pixel that will be filled in and we should work with the partial pixels for these values.

Reply:
Ahh, I see what you are saying. I'll probably leave it as is for now, as it's probably "good enough" for now.